File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(root == nullptr) {
            return nullptr;
        }
        
        //traverse
        TreeNode* tmp = root->left;
        root->left = root->right;
        root->right = tmp;

        //recursive call
        invertTree(root->left);
        invertTree(root->right);

        return root;

    }

    //check invert status
    

    //if we need to invert we invert





};


//Think cases, if we want to invert a binary tree, root stays the same, 
//every left node isnow the right node and the other way around

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import heapq
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        #slightly unoptimal solution and then optimize from there
        h = []
        def inorder(node):
            if not node:
                return
            inorder(node.left)
            heapq.heappush(h,node.val)
            inorder(node.right)
        inorder(root)
        for i in range(k):
            k_th_smallest = heapq.heappop(h)
        return k_th_smallest
        



        